Add yamllint as a lint tool for YAML sources

vkleen opened the pull request:
This PR adds a backend for running [yamllint](https://pypi.org/project/yamllint/) on YAML source files.

The `yamllint` backend is targetless. It finds all YAML files in the repository matching a configurable regular expression and calls `yamllint` on them. The include and exclude regexps and command line arguments are specified as subsystem options.

The found YAML files are partitioned by configuration files. The plugin searches for files named `.yamllint` by default. This filename is configurable by a subsystem option. A found configuration file is used for the entire subtree of files rooted at the directory that contains it. Closer configuration files take priority. Note that `yamllint` is called with its working directory being the repository root. So, any exclude directives for `yamllint` in a configuration file need to be expressed as full paths from the root.

kaos commented on src/python/pants/backend/tools/yamllint/subsystem.py:
```diff
@@ -0,0 +1,49 @@
+        ]
+        return ConfigFilesRequest(
+            specified=self.config,
+            specified_option_name="[{self.options_scope}].config",
```
missing `f"..`

vkleen replied:
:facepalm: Thanks :)

kaos commented on src/python/pants/backend/tools/yamllint/target_types.py:
```diff
@@ -0,0 +1,38 @@
+
+class YamlSourcesGeneratingSourcesField(MultipleSourcesField):
+    uses_source_roots = False
+    default = ("*.yaml",)
```
maybe include `*.yml` in the defaults as well?

kaos reviewed (commented):
This is looking sensible to me, thank you!

Maybe add a test for a badly formatted yaml file as well.

vkleen commented on src/python/pants/backend/tools/yamllint/register.py:
```diff
@@ -0,0 +1,12 @@
+
+
+def target_types():
+    return [YamlSourceTarget, YamlSourcesGeneratorTarget]
```
Do you mean that only the `register.py` should move to experimental, or should all the python modules move there?

kaos replied:
Just the `register.py` file. That way it's easy to graduate it out of experimental in the future, it's just one file with little code that needs to move :)

Eric-Arellano commented on src/python/pants/backend/tools/yamllint/target_types.py:
```diff
@@ -0,0 +1,38 @@
+
+
+class YamlSourceField(SingleSourceField):
+    expected_file_extensions = (".yaml",)
```
```suggestion
    expected_file_extensions = (".yaml", ".yml")
```

Eric-Arellano reviewed (commented):
Thanks! Yamllint support sounds great :)

Thoughts on not adding a target type for this, and instead having it operate on any YAML file? We have less precedent for that because the technology is newer, but it's a direction we're interesting in moving more towards. 

My main concern with targets is the intersection with `files` vs `resources` -- usually you'd use those targets to describe YAML files and add them as `dependencies` to your other targets like `python_sources`. That's going to be necessary to keep doing, and the backends like Python won't do anything when you add a `yaml_source` target to their `dependencies` because they only add `file` and `resource`.

Another benefit of not using targets is it's less boilerplate.

The only thing we don't have totally figured out is how to handle telling Pants to skip linting certain YAML files. Normally, we have the `skip_my_tool` field on individual targets, like `skip_flake8`. That can be nice because it's decentralized. Here, the only idea I have is the option `[yammlint].skip_globs`. It should work out well, only is a difference that it's now a centralized list rather than decentralized in BUILD files.

What do you think? 

Here's an example of a file-less linter, and I'm happy to give better instructions if you agree with this path:

https://github.com/pantsbuild/pants/blob/afd41699a791208b785a893e1687950b2a0f2f36/src/python/pants/backend/project_info/regex_lint.py#L258-L342

kaos commented:
Good point @Eric-Arellano !

> Here, the only idea I have is the option [yammlint].skip_globs. It should work out well, only is a difference that it's now a centralized list rather than decentralized in BUILD files.

I have another idea. That is if we generate synthetic targets for all yaml files, then we have a place for putting the `skip_yamllint=True` for relevant files. We only have to make sure that the synthetic targets don't interfere with the file ownership, which I'm not particularly familiar with how it works (if there's always an issue where multiple targets own the same source file or not for instance).

Note that with synthetic targets, you could easily exclude entire subtrees using `__defaults__` for the synthetic `yaml_file` target.

Eric-Arellano commented:
> That is if we generate synthetic targets for all yaml files

I suspect avoiding targets will be much simpler and ergonomic for end users. Targets are a foreign concept when you first starting using Pants. It doesn't seem like a very big deal to have a centralized ignore list — that's how most tools work anyways with their config, like Black.

kaos commented:
> It should work out well, only is a difference that it's now a centralized list rather than decentralized in BUILD files.

To me, this looks like a hint that this deserves careful consideration. It steps away from how things work "the Pants way" where we keep configuration about the sources Pants work upon in the BUILD files, and the configuration for Pants itself and the tools it uses are in `pants.toml` and other dedicated configuration files (as supported by each respective tool).

My point being, if driven to its edge, is that we could get rid of all the BUILD files and configure everything centrally with globs and knobs, and while we have a desire to reduce the amount of targets needed, ideally to a point where we wouldn't need targets at all, requires carefully looking at how to get there safely.

For completeness I'll present some counter arguments for the above as well. If we consider this glob to be a configuration of how Pants operates, then it would be fine, I guess. We also have the preamble backend with configuration rather than targets, same thing. 

OK, I may have convinced myself that I'm fine with it being a central option.. will leave my reasoning here nonetheless for posterity and for others to consideration to find faults/mistakes in my logic etc. :)

smelc commented:
Hello, I'm the consumer of the yamllint plugin :wave:

I don't have a high-level view of `pants` since I only started introducing pants in my monorepo a few months ago, but here are my thoughts on the discussion above:

> Thoughts on not adding a target type for this, and instead having it operate on any YAML file?

@Eric-Arellano> in the no-target-type scenario, a `BUILD` file would still be required in the folder where yaml files live right?

> Another benefit of not using targets is it's less boilerplate.

In our use case, the boilerplate of using this plugin was only adding `yaml_sources(name="yaml_sources")` stanzas. So it is well accepted. It's not surprising.

> The only thing we don't have totally figured out is how to handle telling Pants to skip linting certain YAML files. Normally, we have the skip_my_tool field on individual targets, like skip_flake8. That can be nice because it's decentralized. Here, the only idea I have is the option [yammlint].skip_globs. It should work out well, only is a difference that it's now a centralized list rather than decentralized in BUILD files.

This sounds a bit surprising to me, because centralizing a configuration piece doesn't scale: there could potentially be frequent conflicts on changes to the centralized file in a large organization. I'm saying this, because I heard this argument in my team when I introduced pants in the monorepo and it seems sensible (even for a monorepo advocate like me :stuck_out_tongue:).

Finally, as a developer introducing pants in my organization, I was expecting that I can write `skip_yamllint`  (as I have seen `skip_tool` examples in a few places already in pants' doc), whereas I was unaware of the `[yammlint].skip_globs` possibility.

Sorry that this doesn't make for a very actionable feedback, but I thought it would be useful to share the view of an average pants user :slightly_smiling_face:

kaos commented:
@smelc Thank you for providing your perspective on this, it is very valuable to us.

>> Thoughts on not adding a target type for this, and instead having it operate on any YAML file?

> @Eric-Arellano> in the no-target-type scenario, a BUILD file would still be required in the folder where yaml files live right?

For target-less also means no need for BUILD file, so your answer here is "no" :) 

> This sounds a bit surprising to me, because centralizing a configuration piece doesn't scale: there could potentially be frequent conflicts on changes to the centralized file in a large organization. I'm saying this, because I heard this argument in my team when I introduced pants in the monorepo and it seems sensible (even for a monorepo advocate like me 😛).

To me, this is tips the scale in favour of having some connection between yaml files and BUILD file targets. 

We'd prefer this to be as transparent as possible however, I think with these attributes:

* `./pants lint ::` should do the right thing by default (linting all yaml files, without requiring them to be owned by any targets) for minimal boilerplate.
* Yaml source files should be able to be owned by `file` / `resource` targets. To still be able to include them in archives/dists/pex'es etc.
* Yaml lint should be able to be skipped using a `skip_yamllint=True` field on such targets. To have a scalable escape hatch.


I believe we can achieve the above with a combination of a target plugin field and careful use of synthetic targets.

> Finally, as a developer introducing pants in my organization, I was expecting that I can write skip_yamllint (as I have seen skip_tool examples in a few places already in pants' doc),

This is also a very important point, and ties in with my comment above with having this "the Pants way". If there are too many different ways we do things, it'll add to the complexity and amount of things to learn as a new user coming to Pants.

>> @kaos (me)
>> OK, I may have convinced myself that I'm fine with it being a central option.. will leave my reasoning here nonetheless for posterity and for others to consideration to find faults/mistakes in my logic etc. :)

So I'm back at thinking that not having this as a central option is preferred.

vkleen commented:
> We'd prefer this to be as transparent as possible however, I think with these attributes:
> 
>     * `./pants lint ::` should do the right thing by default (linting all yaml files, without requiring them to be owned by any targets) for minimal boilerplate.
> 
>     * Yaml source files should be able to be owned by `file` / `resource` targets. To still be able to include them in archives/dists/pex'es etc.
> 
>     * Yaml lint should be able to be skipped using a `skip_yamllint=True` field on such targets. To have a scalable escape hatch.
> 
> 
> I believe we can achieve the above with a combination of a target plugin field and careful use of synthetic targets.

This sounds like the most sensible approach. Do you have an example for how synthetic targets work? I'd like to try my hand at implementing something that gets `./pants lint ::` to work for yaml files without having to specify a `yaml_sources` directive in a BUILD file.

kaos commented:
> > We'd prefer this to be as transparent as possible however, I think with these attributes:
> > ```
> > * `./pants lint ::` should do the right thing by default (linting all yaml files, without requiring them to be owned by any targets) for minimal boilerplate.
> > 
> > * Yaml source files should be able to be owned by `file` / `resource` targets. To still be able to include them in archives/dists/pex'es etc.
> > 
> > * Yaml lint should be able to be skipped using a `skip_yamllint=True` field on such targets. To have a scalable escape hatch.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I believe we can achieve the above with a combination of a target plugin field and careful use of synthetic targets.
> 
> This sounds like the most sensible approach. Do you have an example for how synthetic targets work? I'd like to try my hand at implementing something that gets `./pants lint ::` to work for yaml files without having to specify a `yaml_sources` directive in a BUILD file.

Awesome. Pending docs for synthetic targets, there are some inline docstrings and examples to glean from the sources. Links provided for some examples etc:

https://github.com/pantsbuild/pants/blob/d209d1558202349896fa268b2c0a1b3c5fd8b546/src/python/pants/engine/internals/synthetic_targets.py#L9-L72

https://github.com/pantsbuild/pants/blob/d209d1558202349896fa268b2c0a1b3c5fd8b546/src/python/pants/backend/python/goals/lockfile.py#L277-L313

As well as #17653 

Have a look at them, and feel free to hook me up on Slack for more in-depth questions/discussions (preferably in #development to keep it in the open)

smelc commented:
Thanks for the details @kaos :ok_hand: Final thoughts:

If there is a decentralized escape mechanism (`skip_yamllint=True` :+1:), not requiring a `BUILD` file sounds fine to me. I'm expecting it, because I come from a Bazel background, but unbiased newcomers to `pants` won't expect it.

Making `./pants lint ::` do the right thing by default is a worthwhile objective :slightly_smiling_face:

kaos commented:
> Thanks for the details @kaos 👌 Final thoughts:
> 
> If there is a decentralized escape mechanism (`skip_yamllint=True` 👍), not requiring a `BUILD` file sounds fine to me. I'm expecting it, because I come from a Bazel background, but unbiased newcomers to `pants` won't expect it.

Just noting that the `skip_yamllint` option goes into a BUILD file, but as long as the default is ok, you'll do just fine without. (in other words, it's not until you need the escape hatch that the BUILD file come into play)
 
> Making `./pants lint ::` do the right thing by default is a worthwhile objective 🙂

Yeah, we strive for removing as much boilerplate as possible, and being able to lint yaml files without having to declare targets for them is a big win in that direction.

thejcannon commented:
Just adding my perspective here.

I mostly agree with Eric in that a world with barely any BUILD is one where pants can operate on files as files, and not as synthetic targets. That has the element of least surprise when people are coming to pants.

Black formats files based on globs, pytest tests based on filenames, etc... And these things are configured in a config (and in the file itself).

It may be that synthetic targets get us there as well, but I worry about the complexity and error messages in such a world.

I think this is something for us to discuss as a team though...

kaos commented:
Thanks @thejcannon, agree there's more discussion to be had before settling this.

> pytest tests based on filenames [...]

pytest doesn't work without targets, though..? Or I'm reading your sentence wrong..

> It may be that synthetic targets get us there as well, but I worry about the complexity and error messages in such a world.

Let's explore that. I'm struggling to see how you'd add file specific configuration without targets..

thejcannon commented:
Re: pytest. I meant vanilla pytest, outside of pants.

Re:targets: if the only config is excluding from running that's really overkill when a glob-based approach works well. Certainly for things with complex metadata like tests (which require dependencies) will likely always be a target one-way-or-another, but things that don't (like YAML files)...

vkleen commented:
Specifically for yamllint, I could easily imagine people wanting to use different configuration files for different YAML files. For example, to turn some lints on or off on a file-specific basis.

kaos commented:
> Re: pytest. I meant vanilla pytest, outside of pants.

Ah, gotcha.
 
> Re:targets: if the only config is excluding from running that's really overkill when a glob-based approach works well. Certainly for things with complex metadata like tests (which require dependencies) will likely always be a target one-way-or-another, but things that don't (like YAML files)...

Right. I'd prefer this conversation on a higher level though, where we look at how this fits with more tools/goals on a wider scope and not just for yamllint specifically, to find the proper target(/-less) UX so we end up with something that is coherent.

benjyw commented:
Thanks for this PR @vkleen! And for your patience as we hash out the design ideas.

My two cents: We generally want to move away from explicit BUILD file targets unless they are necessary. We're already down to targets often being just one line, like `python_sources()`, so at that point, why not do away with them entirely... 

Targets may be needed when metadata is required, but even then, the concept of a "target type" may be overspecified. As Eric stated, a .yaml file often needs to be consumed as a source of a `resources()` or `files()` target, and adding it to yaml_sources() target as well creates ambiguity as to which target "owns" the source.

So I'd love to see this implemented without the target type, at least initially. And later, if local config is needed, maybe it can be added to a generic `resources()/files()` target, somehow. E.g., 

```
resources(name='configs', sources=['*.yml'], skip_lint=['yamllint'])
```

Or something...

vkleen commented:
Thank you for your input @benjyw! I've rather enjoyed reading these design discussions and learning more about Pants in the process. I like the idea of a more generic `resources()` target type that could call `yamllint` for YAML files that it happens to own.

Is there a reasonable way to make `yamllint` opt-in in `pants.toml` in this model? Currently, for the `yaml_source()` targets to be available, you need to explicitly import the backend in `pants.toml`. I would hope for something similar here, like importing the `yamllint` backend would magically add this functionality to a pre-existing `resources()` target.

thejcannon commented:
Yup that's totally doable. Plugins can add fields to existing targets.

My 2c is that we ought to do this targetless all the way, since it has the fewest assumptions.

Either way, let us know what you need to continue on. We're here to help! 😀

benjyw commented:
Yes, I agree that we should start targetless, and allow you to opt out specific files via a global config option. If and when that becomes unwieldy, we can consider how to handle it.

smelc commented:
@vkleen> is this ready for review?

vkleen commented:
At the moment, I think it's still doing the wrong thing in the presence of yamllint configurations files. Namely, it picks a single one (and I'm not sure which one) and applies it for all YAML files it finds. I still want to take another look at the partitioning logic, either something like #17739 or maybe just run one invocation of yamllint per YAML file found as a first pass.

So, I don't think it's ready for merging yet but preliminary reviews are always welcome!

thejcannon commented:
In this case I'd expect we do as the Romans do and just use one partition and let the vanilla config finder do it's work.

Then later we can improve it by letting it partition by config.

I would want a perfect implementation to hold up a working one

vkleen commented:
I think this is now ready for review. I've updated the PR description to describe the current implementation. In a nutshell, the plugin is now completely targetless and I've implemented a partitioning scheme by (autodiscovered) configuration files.

vkleen commented:
I've fixed an order dependence in my integration test for the partitioning logic (the CI failure in shard 0). The other CI failure seem to be unrelated to my changes at first glance.

benjyw commented:
@vkleen you might have to merge latest main and push, it looks like no CI jobs were triggered?

thejcannon commented on src/python/pants/backend/tools/yamllint/subsystem.py:
```diff
@@ -0,0 +1,55 @@
+from pants.util.strutil import softwrap
+
+
+class Yamllint(PythonToolBase):
```
We'll want a lockfile for this, check out some of the other Python tools for how that looks

vkleen replied:
I've copied the approach from ClangFormat in https://github.com/pantsbuild/pants/pull/17821/commits/d4ba3ccec669b20d912c0277473b36af6d786ce4.

thejcannon commented on src/python/pants/backend/tools/yamllint/subsystem.py:
```diff
@@ -0,0 +1,55 @@
+        "--config-file-name",
+        default=".yamllint",
+        advanced=True,
+        help=lambda cls: softwrap(
```
You aren't using `cls`, so the lambda is unecessary.

Same below...

thejcannon commented on src/python/pants/backend/tools/yamllint/subsystem.py:
```diff
@@ -0,0 +1,55 @@
+    file_glob_include = StrListOption(
+        "--include",
+        default=["**/*.yml", "**/*.yaml"],
+        help=lambda cls: softwrap(
+            """
+                Glob for which YAML files to lint.
+                """
+        ),
```
This and the one below it don't need to be softwrapped, since they will fit.

thejcannon commented on src/python/pants/backend/tools/yamllint/rules.py:
```diff
@@ -0,0 +1,185 @@
+
+
+@rule(desc="Lint using yamllint", level=LogLevel.DEBUG)
+async def run_yamllint(request: YamllintRequest.Batch[Any, Any], yamllint: Yamllint) -> LintResult:
```
The first generic param should be `str`. The second should be `PartitionInfo`

thejcannon commented on src/python/pants/backend/tools/yamllint/rules.py:
```diff
@@ -0,0 +1,185 @@
+    return Partitions(
+        (
+            *(
+                Partition(tuple(files), PartitionInfo(config_snapshot=config_snapshot))
```
`set`s change order from invocation to invocation, and since you put this in a tuple (ordered) this will result in cache misses for the same contents.

`sorted` is your friend.

(same below)

thejcannon commented on src/python/pants/backend/tools/yamllint/rules.py:
```diff
@@ -0,0 +1,185 @@
+from pants.backend.python.util_rules.pex import Pex, PexProcess, PexRequest
+from pants.backend.tools.yamllint.subsystem import Yamllint
+from pants.core.goals.lint import LintFilesRequest, LintResult
+from pants.core.goals.tailor import group_by_dir
```
It's only a little weird we're importing this from `tailor`...

We should probably put this in a utility file :smile:

thejcannon replied:
Seems like a good place to me :+1:

vkleen replied:
I've made a PR for that at #18029.

thejcannon commented on src/python/pants/backend/tools/yamllint/rules.py:
```diff
@@ -0,0 +1,185 @@
+    source_dir_to_config_files: FrozenDict[str, str]
+
+
+def find_nearest_ancestor_file(files: set[str], dir: str, config_file: str) -> str | None:
+    while True:
+        candidate_config_file_path = os.path.join(dir, config_file)
+        if candidate_config_file_path in files:
+            return candidate_config_file_path
+
+        if dir == "":
+            return None
+        dir = os.path.dirname(dir)
```
You stole this from `scalafmt` :wink: 

Please move it to a utility and both can use it.

vkleen replied:
You caught me :laughing: Would `util/dirutil.py` be a good place for it, or should that go somewhere else?

thejcannon commented on src/python/pants/backend/tools/yamllint/rules.py:
```diff
@@ -0,0 +1,185 @@
+        dir = os.path.dirname(dir)
+
+
+@rule
+async def gather_config_files(
```
Can you add a `@TODO` here (and to `scalafmt`) that these implementations are similar and should likely be genericized and shared? You don't have to address it, just jot it down.

thejcannon reviewed (commented):
LGTM, after this round of comments I think we're golden.

Thanks! Super exciting!

vkleen commented:
Just to make sure, this PR should be merged after #18029. I'll rebase afterwards, since I cherry picked the `group_by_dir` change onto this branch.

thejcannon reviewed (approved):
You've done absolutely amazing work here.

Thanks for sticking it out throughout the changes :smile:

benjyw commented:
@kaos , good to merge?

@vkleen Thanks for this improvement, and for your patience throughout the review.

thejcannon commented:
I also wanna add I think you've pushed us close to the edge for just implementing https://github.com/pantsbuild/pants/issues/17739 :rocket: .

If you would like to take a stab at it we'd love to help you out.

kaos reviewed (approved):
Impressive work!

Would've loved to see a small test for `find_nearest_ancestor_file()`, but can't expect that from this PR as that method isn't introduced here..

Thanks!

smelc commented:
Is there anything @vkleen needs to do? Or will someone push the _merge_ button at some point in the future :sweat_smile:?

thejcannon commented:
Nope :rocket:
